### PR TITLE
Upgrade to Pydantic v2 (GSI 337)

### PR DESCRIPTION
**Pull request by TheByronHimes**
`__get_validators__` is no longer used and should be replaced with `__get_pydantic_core_schema__` according to [the docs](https://docs.pydantic.dev/latest/migration/#defining-custom-types). For this repo, that only affects the `DateTimeUTC` class. 

Running the tests under `examples` raises some deprecation warnings whose root cause is a little unclear but is coming from  `hexkit.inject.AsyncConstructor`. It's moot because that only exists due to the `dependency_injector` framework, which we're moving away from.

The other notable change is with `.schema()`, which is now `.model_json_schema()`:
- The 'type' key is absent from the field information in the schema dump. This seems like it's part of dropping input-type preservation. In the affected test, `tests/unit/httpyexpect/test_server_exceptions::test_http_custom_exception_body`, you can see how this extra information is distilled into simply `{'const': MyCustomHttpException.exception_id}`. 
- Additionally, the top-level `definitions` key is changed to `$defs`.

### CODE REVIEW

File: src/ghga_service_commons/httpyexpect/server/exceptions.py [hunk 1]
```diff
@@ -18,6 +18,8 @@
 
 from abc import ABC
 
+from pydantic import ConfigDict
+
 try:  # workaround for https://github.com/pydantic/pydantic/issues/5821
```
Comment (Cito) [inline src/ghga_service_commons/httpyexpect/server/exceptions.py:23]: This workaround should not be necessary any more, can you check?

File: src/ghga_service_commons/utils/utc_dates.py
Comment (Cito) [inline src/ghga_service_commons/utils/utc_dates.py:1]: Pydantic v2 has [now ](https://github.com/pydantic/pydantic/pull/4812) better support for datetime, and maybe we do not need our own utc date class any more.

We should check/change this separately from this PR though. Can you put it on the refactoring list or create a task for the next sprint?

File: tests/unit/httpyexpect/test_server_exceptions.py [hunk 2]
```diff
@@ -125,21 +125,19 @@ def test_http_custom_exception_body():
     assert issubclass(body_model, pydantic.BaseModel)
 
     # evaluate the schema:
-    body_schema = body_model.schema()
+    body_schema = body_model.model_json_schema()
     assert set(body_schema["properties"].keys()) == {
```
Comment (Cito) [inline tests/unit/httpyexpect/test_server_exceptions.py:129]:
```suggestion
    assert set(body_schema["properties"]) == {
```

**Review by Cito — DISMISSED** [review]
Looks good to me, but see comments.

File: src/ghga_service_commons/httpyexpect/server/exceptions.py (hunk 1 above)
Comment (TheByronHimes) [reply src/ghga_service_commons/httpyexpect/server/exceptions.py]: You are right, it seems to be fixed now.

**Review by KerstenBreuer — COMMENTED** [review]
Looks good, just the points that @Cito raised.

**Comment by KerstenBreuer** [comment]
And coverage is not working. Please have a look at these PRs:
https://github.com/ghga-de/microservice-repository-template/commit/001678ca15d43d92b6c936cdd8a8ff398f1a1f04
https://github.com/ghga-de/microservice-repository-template/commit/ed01f272458937a9c66a9e88e4b8c83e4a6ca12b

File: src/ghga_service_commons/utils/utc_dates.py
Comment (TheByronHimes) [reply src/ghga_service_commons/utils/utc_dates.py:1]: @Cito I made an item to look into it

**Review by Cito — APPROVED** [review]

**Review by KerstenBreuer — APPROVED** [review]

**Pull request merged**